iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java

```diff
@@ -4,151 +4,116 @@
-                    return buildGaugeWithoutLabels(metricName, help, value);
-                })
-                .collect(Collectors.toList());
+    private static List<String> labelNames(List<ExtractedMetric> extractedMetrics) {
+        return firstElementFromSingletonListOrThrow(extractedMetrics, ExtractedMetric::getLabelNames);
     }
 
-    private static String durationMetricName(ReportEntry reportEntry) {
-        return String.format("%s_seconds",
-                reportEntry.getKey().toString().replace(REPORT_ENTRY_KEY_SEP, METRIC_NAME_SEP));
+    private static List<List<String>> labelValues(List<ExtractedMetric> extractedMetrics) {
+        return extractedMetrics.stream().map(ExtractedMetric::getLabelValues).collect(Collectors.toList());
     }
 
-    private static Double millisecondsToSeconds(Double milliseconds) {
-        return milliseconds / 1000;
+    private static List<Double> values(List<ExtractedMetric> extractedMetrics) {
+        return extractedMetrics.stream().map(ExtractedMetric::getValue).collect(Collectors.toList());
     }
 
-    private static Gauge buildGaugeWithLabels(String metricName,
-                                              String help,
-                                              String[] labelNames,
-                                              List<MetricNameAndLabelValuesAndValue> metricNameAndLabelValuesAndValueList) {
-        Gauge gauge = Gauge.build()
-                .name(metricName)
-                .help(help)
-                .labelNames(labelNames)
-                .create();
-        metricNameAndLabelValuesAndValueList.forEach(x -> gauge.labels(x.labelValues).set(x.value));
-        return gauge;
+    private static Stream<Gauge> convertDurationReportEntries(Stream<ReportEntry> reportEntries,
+                                                              String help,
+                                                              DurationReportEntryMetricExtractor.Extractor metricExtractor,
+                                                              GaugesBuilder.BuilderWithoutLabels gaugesWithoutLabelsBuilder) {
+        return reportEntries
+                .map(metricExtractor::extract)
+                .map(extractedMetric -> gaugesWithoutLabelsBuilder.build(extractedMetric.getMetricName(), help, extractedMetric.getValue()));
     }
 
-    private static Gauge buildGaugeWithoutLabels(String metricName,
-                                                 String help,
-                                                 double value) {
-        Gauge gauge = Gauge.build()
-                .name(metricName)
-                .help(help)
-                .create();
-        gauge.set(value);
-        return gauge;
+    private static <E, X> X firstElementFromSingletonListOrThrow(List<E> list, Function<E, X> mapper) {
+        List<X> collect = list.stream().map(mapper).distinct().collect(Collectors.toList());
```

Comment:
Hmm, do we have to make stream out of the list to take the first element?

Reply:
Generally no, but in this case, apart from applying also a mapper, we want to check that this the list contains only one unique element.

Reply:
To make this more visible I've changed the name of this method to `uniqueElementOrThrow`.